spatialos-sdk/src/worker/schema.rs

```diff
@@ -41,3 +109,151 @@ pub trait SchemaPrimitiveField {
+/// types, and provides a blanket implementation of the [`Field`] trait so that the
+/// types can be serialized as part of larger objects.
+///
+/// You should generally not have to implement this trait manually for any types.
+/// The implementation for any schema-defined types will be generated for you by the
+/// code generator provided with the SDK.
+///
+/// [components]: https://docs.improbable.io/reference/14.2/shared/schema/reference#components
+/// [types]: https://docs.improbable.io/reference/14.2/shared/schema/reference#types
+/// [`SchemaObject`]: struct.SchemaObject.html
+/// [`Field`]: trait.Field.html
+pub trait ObjectField: Sized + Clone {
+    /// Deserializes an instance of this type from a [`SchemaObject`].
+    ///
+    /// Note that deserialization can never fail as the underlying SpatialOS API ensures
+    /// that missing or invalid fields will always provide a valid default value during
+    /// deserialization. As such, if the data in `object` doesn't match what is expected
+    /// by the deserialization logic (e.g. due to a version mismatch between the current
+    /// schema version and the version `object` was generated from), the returned object
+    /// will have default values for any fields that are missing in `object` or have a
+    /// type differing from what is expected.
+    ///
+    /// [`SchemaObject`]: struct.SchemaObject.html
+    fn from_object(object: &SchemaObject) -> Self;
+
+    /// Serializes an instance of this type into a [`SchemaObject`].
+    ///
+    /// Note that serialization can never fail, as there is no way to form an invalid
+    /// instance of a schema object type.
+    ///
+    /// [`SchemaObject`]: struct.SchemaObject.html
+    fn into_object(&self, object: &mut SchemaObject);
+}
+
+impl<T> Field for T
+where
+    T: ObjectField,
+{
+    type RustType = Self;
+
+    fn get_or_default(object: &SchemaObject, field: FieldId) -> Self::RustType {
+        Self::from_object(object.get_object(field))
+    }
+
+    fn index(object: &SchemaObject, field: FieldId, index: usize) -> Self::RustType {
+        Self::from_object(object.index_object(field, index))
+    }
+
+    fn count(object: &SchemaObject, field: FieldId) -> usize {
+        object.object_count(field)
+    }
+
+    fn add(object: &mut SchemaObject, field: FieldId, value: &Self::RustType) {
+        value.into_object(object.add_object(field));
+    }
+}
+
+/// The error type returned when a conversion to a schema enum type fails.
+#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct UnknownDiscriminantError;
+
+/// A type that corresponds to a [schemalang enumeration][schema-enum].
+///
+/// Schemalang enums are C-like enums (i.e. they only have a discriminant and do not
+/// contain any additional data) represented as `u32` values when serialized. They
+/// default to the first variant defined in their schema declaration.
```

Comment:
What do you mean by "default to the first variant"? Who is defaulting what here?

Reply:
I guess the `Default` impl is technically determined by code generation, though I was operating under the assumption that trying to get a non-existent field from a schema object would return a default value of 0, which would line up with the first variant of the enum. As I type this out, I realize that might not actually be a valid assumption 🤔

Reply:
Sadly schema enums don't always start at 0 :( 

```
enum MyEnum 
{
	VALUE = 1;
}
```

is perfectly valid schema. This also means that getting a value of `1` for a field of type MyEnum maps to `VALUE` and a value of `0` is 🤷‍♂ 

https://github.com/spatialos/gdk-for-unity/issues/960 :wink: 


---

I'm not sure exactly what the behaviour of calling `Schema_GetUint32(obj, 2);` where field 2 doesn't exist. Paging @dgavedissian!

Reply:
Ah, I was just going to ask how this is handled in other GDKs! Seems like the answer is "it isn't" 😭

Reply:
Yeah. Enums aren't really a thing in the `Schema_` serialization layer. If you call `Schema_GetUint32(obj, 2)`, then it'll just return 0 if field ID 2 doesn't exist as Jamie said. I suggest that in codegen you special case the enum field in a way like this:

```
if Schema_GetUint32Count(obj, 2) == 0 {
  MyEnum::default()
} else {
  Schema_GetUint32(obj, 2)
```

Reply:
That's effectively what the current implementation in the PR is doing: If the value doesn't match a known variant, you get the default value out. I'm wondering if that's the right thing to do, though. The user has no way to specify which variant is the default, and in some cases there might not be a meaningful default. Silently spitting out potentially invalid data seems like the wrong thing to do, at least for a Rust API.

Schemalang [supports custom annotations](https://docs.improbable.io/reference/14.2/shared/schema/reference#annotations), so I suppose what we can do is default to producing an error if the value is unknown, but allow the user to specify a default value to use, which would allow serialization to succeed in the face of an unknown variant.